test/api/model/web_auth_test.dart

```diff
@@ -0,0 +1,71 @@
+import 'package:checks/checks.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:zulip/api/model/web_auth.dart';
+
+import '../../example_data.dart' as eg;
+
+void main() {
+  group('WebAuthPayload', () {
+    const otp = '186f6d085a5621ebaf1ccfc05033e8acba57dae03f061705ac1e58c402c30a31';
+    final encryptedApiKey = debugEncodeApiKey(eg.selfAccount.apiKey, otp);
+    final wellFormed = Uri.parse(
+      'zulip://login?otp_encrypted_api_key=$encryptedApiKey&email=self%40example&user_id=1&realm=https%3A%2F%2Fchat.example%2F');
+
+    test('basic happy case', () {
+      final result = WebAuthPayload.parse(wellFormed);
+      check(result)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.equals(1)
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(result.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    // TODO(server-5) remove
+    test('legacy: no userId', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('user_id');
+      final result = WebAuthPayload.parse(
+        wellFormed.replace(queryParameters: queryParams));
+      check(result)
+        ..otpEncryptedApiKey.equals(encryptedApiKey)
+        ..email.equals('self@example')
+        ..userId.isNull()
+        ..realm.equals(Uri.parse('https://chat.example/'));
+      check(result.decodeApiKey(otp)).equals(eg.selfAccount.apiKey);
+    });
+
+    test('parse fails when an expected field is missing', () {
+      final queryParams = {...wellFormed.queryParameters}..remove('email');
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when otp_encrypted_api_key is nonsense', () {
+      final queryParams = {...wellFormed.queryParameters}..['otp_encrypted_api_key'] = 'asdf';
+      final input = wellFormed.replace(queryParameters: queryParams);
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when host is not "login"', () {
+      final input = wellFormed.replace(host: 'foo');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('parse fails when scheme is not "zulip"', () {
+      final input = wellFormed.replace(scheme: 'https');
+      check(() => WebAuthPayload.parse(input)).throws<FormatException>();
+    });
+
+    test('decodeApiKey fails when otp is nonsense', () {
```

Comment:
nit:
```suggestion
    test('decodeApiKey fails when otp is wrong length', () {
```

Seems more precise about what's really being checked here. (And an OTP is random noise when it *is* properly generated, so the reader may wonder what it means for it to be "nonsense".)